### PR TITLE
[concept] add CLI for interactive debugging / learning

### PR DESCRIPTION
**Pull request by agavra**
This PR introduces a REPL for SlateDB, which is mostly intended as an entrypoint for developers like me to be able to play around, set breakpoints to explore the SlateDB codebase interactively. Eventually we can add all sorts of admin commands to this to test out specific scenarios (such as triggering compaction, etc...) and manually explore situations.

In my work with previous databases I found this to be a really helpful feature to have. Secretly a motivation for me was to just learn rust/explore SlateDB so feel free to disagree that this is something we should have!

Here's an example run:
```
$ cargo run -p cli
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/cli`
Welcome to the SlateDB REPL!
WARNING: This is currently intended for development use.

  ____  _       _       ____  ____
 / ___|| | __ _| |_ ___|  _ \| __ )
 \___ \| |/ _` | __/ _ \ | | |  _ \
  ___) | | (_| | ||  __/ |_| | |_) |
 |____/|_|\__,_|\__\___|____/|____/

Type 'help' to see available commands.

>> put key 'a complex value'
Successfully inserted key-value pair.
>> put key2 another_value
Successfully inserted key-value pair.
>> get key
a complex value
>> get key2
another_value
>>
CTRL-C
Exiting REPL. Goodbye!
```

### CODE REVIEW

**Comment by palash25** [comment]
this looks pretty cool just FYI you need to run `cargo fmt` to make the CI green and there is already an issue about a CLI tool that someone else is working on https://github.com/slatedb/slatedb/issues/151

slightly different from your idea but I feel like the issue might be a superset of what you are proposing, so you might want to sync with the maintainers and the contributor working on that one.

**Comment by agavra** [comment]
Thanks @palash25! I wasn't aware of the existing ticket, but Chris brought up the same comment on the Discord and we were chatting about it there. 

To summarize the discussion there: I think it makes sense to keep both a REPL and a CLI, REPLs are helpful for debugging/interactive/learning purposes and CLIs are really helpful from a devops perspective (I want to be able to integrate my tooling and CI/CD with a CLI, I don't want to deal with a REPL).

As far as the specific commands, #151 is definitely a superset of what I have proposed here. Either way, I'll sync up with @Hrushi20 to see what the status is on their work and can incorporate whatever they have into this 😄

File: cli/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+            object_store_cache_options: ObjectStoreCacheOptions::default(),
+        };
+        let db = Db::open_with_opts(Path::from(storage_path), options, object_store).await?;
+        Ok(SlateDbREPL { db })
+    }
+
+    async fn put(&self, key: &str, value: &str) -> Result<(), Box<dyn std::error::Error>> {
+        self.db.put(key.as_bytes(), value.as_bytes()).await;
+        Ok(())
+    }
```
Comment (palash25) [inline cli/src/main.rs:50]: `self.db.put` doesn't return any error so this can be simplified to just one line

```suggestion
    async fn put(&self, key: &str, value: &str) {
        self.db.put(key.as_bytes(), value.as_bytes()).await
    }
```

in case this was done just to have a consistent return type in all the API methods then i guess its ok, I am very new to Rust so I don't have strong opinions about it just pointing it out

File: cli/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    async fn get(&self, key: &str) -> Result<Option<String>, Box<dyn std::error::Error>> {
+        if let Some(bytes) = self.db.get(key.as_bytes()).await? {
+            let value = String::from_utf8(bytes.to_vec())?;
+            Ok(Some(value))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn delete(&self, key: &str) -> Result<(), Box<dyn std::error::Error>> {
```
Comment (palash25) [inline cli/src/main.rs:61]: same comment as the one on `put` method

File: cli/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            l0_sst_size_bytes: 128,
+            l0_max_ssts: 8,
+            max_unflushed_memtable: 2,
+            compactor_options: Some(CompactorOptions::default()),
+            compression_codec: None,
+            block_cache_options: Some(InMemoryCacheOptions::default()),
+            object_store_cache_options: ObjectStoreCacheOptions::default(),
+        };
+        let db = Db::open_with_opts(Path::from(storage_path), options, object_store).await?;
+        Ok(SlateDbREPL { db })
```
Comment (palash25) [inline cli/src/main.rs:44]: nitpick

```suggestion
        Ok(Self { db })
```

**Review by palash25 — COMMENTED** [review]
just skimmed through the code, I am new to rust so mostly pointed out cosmetic changes i don't see any big issues with the PR, looks good

File: cli/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                let trimmed = line.trim();
+                if trimmed.is_empty() {
+                    continue;
+                }
+                match rl.add_history_entry(trimmed) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        eprintln!("Could not save line to history: {}", e)
+                    }
+                }
```
Comment (palash25) [inline cli/src/main.rs:123]: since we don't care about the `Ok` case we can re-write this as `if let`

```suggestion
                if let Err(e) rl.add_history_entry(trimmed) {
                    eprintln!("Could not save line to history: {}", e)
                }
```

File: cli/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+                    Ok(p) => p,
+                    Err(e) => {
+                        eprintln!("Error parsing input: {}", e);
+                        continue;
+                    }
+                };
+
+                if parts.is_empty() {
+                    continue;
+                }
```
Comment (palash25) [inline cli/src/main.rs:137]: can this snippet be moved up and the continue combined with the `trimmed.is_empty()`? and then we can check for history entry and we can remove the `parts_result` intermediate variable as its used only once

like this

```rust
let trimmed = line.trim();
let parts = match split(trimmed) {...}
if trimmed.is_empty() || parts.is_empty() {
       continue;
}

// then rl.add_history_entry(trimmed) .......
```

seems a bit cleaner to me if it doesn't break any logic

File: cli/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+                break;
+            }
+            Err(ReadlineError::Eof) => {
+                println!("CTRL-D");
+                println!("Exiting REPL. Goodbye!");
+                if let Err(e) = db_cli.close().await {
+                    eprintln!("Error closing the database: {}", e);
+                }
+                break;
```
Comment (palash25) [inline cli/src/main.rs:211]: this snippet is used in 3 instances, maybe we can move this to a separate function

File: cli/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                            Ok(Some(value)) => println!("{}", value),
+                            Ok(None) => println!("Key not found."),
+                            Err(e) => eprintln!("Error executing get: {}", e),
+                        }
+                    }
+                    "delete" => {
+                        if parts.len() != 2 {
+                            println!("Usage: delete <key>");
+                            continue;
+                        }
```
Comment (palash25) [inline cli/src/main.rs:170]: nitpick: similar code is repeated thrice, we can have a function like `validate_args(parts, valid_len, usage_message)` defined somewhere and then call it with in all the command cases

File: cli/src/main.rs (hunk 4 above)
Comment (agavra) [reply cli/src/main.rs:123]: Nice! Didn't know this about rust syntax (this is also my very first Rust PR) -- I assumed it must be possible

File: cli/src/main.rs (hunk 6 above)
Comment (agavra) [reply cli/src/main.rs:211]: Actually I just put it at when exiting the loop instead of in the method. Thanks for the tip!

File: cli/src/main.rs (hunk 7 above)
Comment (agavra) [reply cli/src/main.rs]: I went ahead and tried to refactor things so that each command go its own implementation of a newly introduced `CliCommand` trait so that we can standardize this a little bit. This will also help with making the REPL and the CLI share more code when it gets to that.

LMKWYT

**Comment by criccomini** [comment]
I took a _brief_ look at the code. This is quite fun!

I noticed that the code uses an `InMemory` object store right now. It seems like it's more geared for playing with the APIs than doing anything operational (like reading/writing KV pairs from S3, dumping manifest content, and so on). Pretty much as you described. As such, I'm leaning toward just continuing the work on #151. @agavra if @Hrushi20 doesn't respond next week, I think you are free to take over that work if you'd like to. 

I think what's instructive about this tool is the actual building of it, rather than using it. I wonder if there's some way we could do something like [rust-playground](https://github.com/rust-lang/rust-playground) or [pglite.dev](https://pglite.dev/)'s in-browser pglite. A nicely themed WASM sandbox that gives developers SlateDB in the browser might be a great way for others to experiment with the API easily.

I'm just spitballing here.

**Comment by criccomini** [comment]
Hrm just noticed that rust-playground requires a Docker backend. That's too much overhead to run for the slatedb.io site. 🤔 Was hoping it was WASM.

**Comment by agavra** [comment]
@criccomini the intention of this specific PR was to have a playground for interacting with the API, but I did imagine that it will grow to be a tool that would help for admin APIs as well (a la #151) -- specifically I was curious to take on #168 and triggering compaction manually from a playground would be helpful (which is exactly where @Hrushi20 was going to start, so I'm holding off additional work on that until we hear back).

The bulk of the work is agnostic to the actual commands being run and the instantiation of the DB 😄

**Comment by criccomini** [comment]
Gotcha. Are you envisioning that all commands could be run either via single-line CLI or via the REPL? I'm just trying to get a grasp on why we need both of these things. For the ops cases, I feel the single-line CLI is more natural.

**Comment by agavra** [comment]
Closing this PR as "concept only" after the discussion here: https://discord.com/channels/1232385660460204122/1232385660946616433/1284307699076436031

TL;DR it probably doesn't make sense to have a REPL interact with slateDB from a production setting and having a unified CLI/REPL would make it challenging to support long-lived operations.

**Comment by Hrushi20** [comment]
Hey! I'm on a vacation. Please feel free to carry on. I haven't made much progress. Thanks.

**Comment by criccomini** [comment]
@Hrushi20 Thanks for the update. Enjoy the vacation. 🏖️
